Extension/src/api.ts

```diff
@@ -0,0 +1,24 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All Rights Reserved.
+ * See 'LICENSE' in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+'use strict';
+
+import * as vscode from 'vscode';
+
+export interface CppToolsApi {
+    registerCustomConfigurationProvider(provider: CustomConfigurationProvider): void;
+}
+
+export interface SourceFileConfiguration {
+    includePaths: string[];    // This must also include the system include path (compiler defaults)
+    defines: string[];        // This must also include the compiler default defines (_MSC_VER, __cplusplus, etc)
+    intelliSenseMode: string; // Probably just ‘msvc-x64’ for Windows
+    forcedInclude?: string[]; // Any files that need to be included before the source file is parsed
+   standard: string;         // This is the C or C++ standard (language server will determine if the file is C or C++ based on this)
+}
+ 
+export interface CustomConfigurationProvider {
+    name: string;
+    provideConfiguration(uri: vscode.Uri, callback: (config: SourceFileConfiguration) => any): void;

```

Comment:
Making this async may result in vscode messages being sent out of order which will result in undefined behavior and confuse the language server.  If you can't synchronize here because you think that your out of proc server is going to take too long to service the request, then we'll need to rethink this. Regardless of what we do here, sendMessage needs to be called before this method returns so you might need to block anyway unless we decide to let the language server open the file with the wrong config until your extension comes back with the correct one.

We probably need some kind of timeout and/or a try/catch block in here to ensure the safety of the provider call (in case a bad extension hangs or crashes)